inc/class-tmsc-sync.php

```diff
@@ -308,45 +324,58 @@ public function complete_sync() {
+
 		global $wpdb;
-		$post_processing_data = $wpdb->get_results(
-			$wpdb->prepare( "SELECT post_id, meta_value from {$wpdb->postmeta} WHERE meta_key = %s", 'tmsc_post_processing' )
+
+		$batch_size = 200;
+		$batch_index = 0;
+		$batch_number = 1;
+		$total_count = $wpdb->get_row(
+			$wpdb->prepare( "SELECT COUNT(*) AS total FROM {$wpdb->postmeta} WHERE meta_key = %s", 'tmsc_post_processing' )
 		);
-		foreach ( $post_processing_data as $row ) {
-			$post_id = $row->post_id;
-			$processor_type = get_post_meta( $post_id, 'tmsc_processor_type' ,true );
-			if ( ! empty( $processor_type ) ) {
-				$relationship_map = apply_filters( "tmsc_{$processor_type}_relationship_map", array() );
-				$data = maybe_unserialize( $row->meta_value );
-
-				foreach ( $data as $key => $ids ) {
-					if ( ! empty( $ids ) && ! empty( $relationship_map[ $key ]['type'] ) ) {
-						if ( 'post' === $relationship_map[ $key ]['type'] ) {
-							$related_posts = get_posts( array(
-								'fields' => 'ids',
-								'suppress_filters' => false,
-								'ignore_sticky_posts' => true,
-								'no_found_rows' => true,
-								'post_type' => $relationship_map[ $key ]['slug'],
-								'post_status' => 'publish',
-								'meta_query' => array(
-									array(
-										'key'     => 'tmsc_legacy_id',
-										'value'   => $ids,
-										'compare' => 'IN',
-									),
-								),
-							) );
-
-							if ( ! empty( $related_posts ) ) {
-								update_post_meta( $post_id, $key, $related_posts );
-							}
-						} elseif ( 'constituent' === $relationship_map[ $key ]['type'] ) {
-							foreach ( $ids as $constituent_type_slug => $role_data ) {
-								$meta_data = array();
-								foreach( $role_data as $role_slug => $constituent_ids ) {
-									if ( ! empty( $constituent_ids ) ) {
-										$related_constituents = get_posts( array(
+
+		if ( ! empty( $total_count->total ) ) {
+			if ( $cli ) {
+				\WP_CLI::line( sprintf( '%s posts need processing...', $total_count->total ) );
+			}
+
+			do {
+
+				$post_processing_data = $wpdb->get_results(
+					$wpdb->prepare(
+						"SELECT post_id, meta_value FROM {$wpdb->postmeta} WHERE meta_key = %s AND post_id = 436501 ORDER BY post_id LIMIT %d, %d",
+						'tmsc_post_processing',
+						$batch_index,
+						$batch_size
+					)
+				);
+
+				if ( ! empty( $post_processing_data ) ) {
+
+					$batch_position = ( $batch_number * $batch_size ) - ( $batch_size - 1 );
+					if ( $cli ) {
+						\WP_CLI::line( sprintf( 'Beginning a batch of %s, post %s of %s', $batch_size, $batch_position, $total_count->total ) );
+					}
+					foreach ( $post_processing_data as $row ) {
+						$post_id = $row->post_id;
+						$processor_type = get_post_meta( $post_id, 'tmsc_processor_type' ,true );
+
+						if ( ! empty( $processor_type ) ) {
+							$relationship_map = apply_filters( "tmsc_{$processor_type}_relationship_map", array() );
+							$data = maybe_unserialize( $row->meta_value );
+							foreach ( $data as $key => $ids ) {
```

Comment:
Do you need to ensure ` ! empty( $data ) && is_array( $data )` before looping through `$data`?